### ISSUE TITLE
Regarding performance of different read dataset methods

### ISSUE THREAD

**Issue by panfengfeng** [uber/petastorm#379]
I use make_petastorm_dataset API (make_petastorm_dataset(reader)) to scan the dataset, I found that it is much worse than reader loop method (for sample in reader), do you have any ideas about the make_petastorm_dataset API

**Comment by selitvin** [comment]
Can you please provide a little bit more information about your setup?
Are you using `make_reader` or `make_batch_reader` functions to create your reader?
Can you show a code snippet that represents how do you do the scanning?

**Comment by panfengfeng** [comment]
` with make_reader(dataset_url, num_epochs=FLAGS.num_epochs, cache_properties=cache_properties, workers_count=FLAGS.workers_count) as train_reader:
    i = 0
    start = time.time()
    for schema_view in train_reader:
      cur_image = schema_view.image
      for image in cur_image:
        i += 1
        if i % FLAGS.record_num == 0:
          end = time.time()
          print("time is " + str(end - start))
          start = end
    print("Finish! the number is " + str(i))`


` with make_reader(dataset_url, num_epochs=FLAGS.num_epochs, cache_properties=cache_properties, workers_count=FLAGS.workers_count) as train_reader:
    train_dataset = make_dataset(train_reader) \
      .apply(tf.data.experimental.unbatch())

    train_dataset = train_dataset.shuffle(buffer_size=3 * FLAGS.batch_size)

    # train_dataset = train_dataset.repeat(FLAGS.num_epochs)

    train_dataset = train_dataset.apply(tf.data.experimental.map_and_batch(parse_record,
                                                                           num_parallel_calls=FLAGS.workers_count,
                                                                           batch_size=FLAGS.batch_size))

    train_iterator = train_dataset.make_one_shot_iterator()

    train_tensor = train_iterator.get_next()

    i = 0
    with tf.Session() as sess:
      sess.run([
        tf.local_variables_initializer(),
        tf.global_variables_initializer(),
      ])

      start = time.time()
      try:
        while True:
          cur_image, cur_label = sess.run(train_tensor)

          for image in cur_image:
            i += 1
            if i % FLAGS.record_num == 0:
              end = time.time()
              print("time is " + str(end - start))
              start = end

      except tf.errors.OutOfRangeError:
        print("Finish! the number is " + str(i))`

one is just using for-loop to scan the dataset, the other is using dataset to scan dataset, (imageNet dataset is 24GB ), and run 2 epochs. 

for-loop takes 36s, 20s, 
dataset takes 58s, 57s, 
tfrecords,takes 41s, 19s

you see that second epoch of dataset take 57s, larger than for-loop and tfrecords
